### PR TITLE
Implement dialect translators

### PR DESCRIPTION
**Pull request by andicuko**

### CODE REVIEW

File: src/dialect_translation/mod.rs [hunk 1]
```diff
@@ -0,0 +1,700 @@
...
+            }
+        }
+    }
+}
+
+/// Trait constructor for building dialect dependent AST parts with default implementations.
+/// We use macros to reduce code repetition in generating functions to convert each Epression variant.
+macro_rules! into_dialect_tranlator_trait_constructor {
+    () => {
+        pub trait IntoDialectTranslator {
```
Comment (ngrislain) [inline src/dialect_translation/mod.rs:128]: FromRelationTranslator

File: src/dialect_translation/mod.rs [hunk 2]
```diff
@@ -0,0 +1,700 @@
...
+//                 _ => $default
+//             }
+//         }
+//     }
+// }
+
+/// Build Sarus Relatioin from dialect speciific AST
+// macro_rules! into_ralation_tranlator_trait_constructor {
+//     () => {
+pub trait IntoRelationTranslator {
```
Comment (ngrislain) [inline src/dialect_translation/mod.rs:545]: FromQueryTranslator

File: src/dialect_translation/mod.rs [hunk 3]
```diff
@@ -0,0 +1,700 @@
...
+                match expr {
+                    expr::Expr::Column(ident) => self.from_column(ident),
+                    expr::Expr::Value(value) => self.from_value(value),
+                    expr::Expr::Function(func) => self.from_function(func),
+                    expr::Expr::Aggregate(agg) => self.from_aggregate(agg),
+                    expr::Expr::Struct(_) => todo!(),
+                }
+            }
+
+            fn from_column(&self, ident: &expr::Identifier) -> ast::Expr {
```
Comment (ngrislain) [inline src/dialect_translation/mod.rs:325]: column

File: src/dialect_translation/mod.rs [hunk 4]
```diff
@@ -0,0 +1,700 @@
...
+            }
+            fn from_cast_as_boolean(&self, expr: &expr::Expr) -> ast::Expr {
+                let ast_expr = self.expr(expr);
+                cast_builder(ast_expr, ast::DataType::Boolean)
+            }
+            fn from_cast_as_date_time(&self, expr: &expr::Expr) -> ast::Expr {
+                let ast_expr = self.expr(expr);
+                cast_builder(ast_expr, ast::DataType::Datetime(Some(64)))
+            }
+            fn from_case(&self, exprs: Vec<&expr::Expr>) -> ast::Expr {
```
Comment (ngrislain) [inline src/dialect_translation/mod.rs:496]: remove all froms

File: src/dialect_translation/mod.rs [hunk 5]
```diff
@@ -0,0 +1,700 @@
...
+/// Build Sarus Relatioin from dialect speciific AST
+// macro_rules! into_ralation_tranlator_trait_constructor {
+//     () => {
+pub trait IntoRelationTranslator {
+    type D: Dialect;
+
+    fn dialect(&self) -> Self::D;
+
+    // It converts ast Expressions to sarus expressions
+    fn try_from_expr(
```
Comment (ngrislain) [inline src/dialect_translation/mod.rs:551]: try_expr

File: src/dialect_translation/mod.rs [hunk 6]
```diff
@@ -0,0 +1,700 @@
...
+        paste! {
+            match $func {
+                // expand arms for binary_op
+                $(
+                    expr::function::Function::$binary_op => binary_op_builder($self.expr($args[0]), ast::BinaryOperator::$binary_op, $self.expr($args[1])),
+                )*
+
+                // expand arms for nullay
+                $(
+                    expr::function::Function::$nullay => $self.[<from_ $nullay:snake>](),
```
[the remark this answers is not in the document]
Comment (andicuko) [reply src/dialect_translation/mod.rs]: corrected.

File: src/dialect_translation/mod.rs (hunk 1 above)
Comment (andicuko) [reply src/dialect_translation/mod.rs]: changed

File: src/dialect_translation/mod.rs (hunk 2 above)
Comment (andicuko) [reply src/dialect_translation/mod.rs]: changed

File: src/dialect_translation/mod.rs (hunk 3 above)
Comment (andicuko) [reply src/dialect_translation/mod.rs]: changed

File: src/dialect_translation/mod.rs (hunk 4 above)
Comment (andicuko) [reply src/dialect_translation/mod.rs]: removed

File: src/dialect_translation/mod.rs (hunk 5 above)
Comment (andicuko) [reply src/dialect_translation/mod.rs]: corrected

**Review by ngrislain — APPROVED** [review]

File: src/display/mod.rs [hunk 7]
```diff
@@ -78,7 +78,11 @@ impl Dot for Relation {
         Command::new("open")
             .arg(format!("/tmp/{name}.html"))
             .output()
-            .expect("Error: this works on MacOS");
+            .or(
+                Command::new("xdg-open")
+                .arg(format!("http://localhost:8000/{name}.html"))
```
Comment (ngrislain) [inline src/display/mod.rs:83]: This will not work, remove

File: src/display/mod.rs [hunk 8]
```diff
@@ -97,7 +101,11 @@ impl Dot for WithContext<&Expr, DataType> {
...
```
Comment (ngrislain) [inline src/display/mod.rs:106]: Why did you do that, please remove everywhere.

File: src/relation/mod.rs [hunk 9]
```diff
@@ -1574,6 +1574,7 @@ impl Ready<Relation> for ValuesBuilder {
 mod tests {
     use super::{schema::Schema, *};
     use crate::{builder::With, data_type::DataType, display::Dot};
+    use crate::{ast};
```
Comment (ngrislain) [inline src/relation/mod.rs:1577]: Why have the {} ?
Can you run cargo fmt?

File: src/sql/relation.rs [hunk 10]
```diff
@@ -43,11 +44,15 @@ This is done in the query_names module.
 /// The Hierarchy of Relations is the context in which the query is converted, typically the list of tables with their Path
 /// The QueryNames is the map of sub-query referrenced by their names, so that links can be unfolded
 #[derive(Clone, Debug, Hash, PartialEq, Eq)]
-struct TryIntoRelationVisitor<'a>(&'a Hierarchy<Arc<Relation>>, QueryNames<'a>);
+pub struct TryIntoRelationVisitor<'a, T: QueryToRelationTranslator + Copy + Clone> {
```
Comment (ngrislain) [inline src/sql/relation.rs:47]: Why is this public?

File: src/display/mod.rs (hunk 7 above)
Comment (andicuko) [reply src/display/mod.rs:83]: This is needed to visualise the relations and run some tests individually when the repository is in a remote linux machine. it requires to run a html server in the /tmp/ directory. It is not general but it allows me to run some individual tests in my development environment.

File: src/sql/relation.rs (hunk 10 above)
Comment (andicuko) [reply src/sql/relation.rs]: It is an error. I reverted it

File: src/display/mod.rs (hunk 8 above)
Comment (andicuko) [reply src/display/mod.rs]: I have reverted it and created the issue [here](https://github.com/Qrlew/qrlew/issues/229)

**Review by ngrislain — APPROVED** [review]

**Pull request merged**